### PR TITLE
Add support for the merge

### PR DESCRIPTION
**Pull request by fvictorio**

### CODE REVIEW

File: .changeset/green-lemons-design.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"hardhat": patch
```
Comment (alcuadrado) [inline .changeset/green-lemons-design.md:2]:
```suggestion
"hardhat": minor
```

File: docs/src/content/hardhat-network/docs/reference/index.md [hunk 2]
```diff
@@ -61,7 +62,7 @@ The block gas limit to use in Hardhat Network's blockchain. Default value: `30_0
 
 #### `hardfork`
 
-This setting changes how Hardhat Network works, to mimic Ethereum's mainnet at a given hardfork. It must be one of `"byzantium"`, `"constantinople"`, `"petersburg"`, `"istanbul"`, `"muirGlacier"`, `"berlin"`, `"london"` and `"arrowGlacier"`. Default value: `"arrowGlacier"`
+This setting changes how Hardhat Network works, to mimic Ethereum's mainnet at a given hardfork. It must be one of `"byzantium"`, `"constantinople"`, `"petersburg"`, `"istanbul"`, `"muirGlacier"`, `"berlin"`, `"london"`, `"arrowGlacier"` and `"merge"`. Default value: `"arrowGlacier"`
```
Comment (alcuadrado) [inline docs/src/content/hardhat-network/docs/reference/index.md:65]: Aren't we missing one here? The latest HF?

Also, the new default should be `merge`.

File: packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts [hunk 3]
```diff
@@ -1365,6 +1365,12 @@ export class EthModule {
       return 0n;
     }
 
+    if (oldBlockTag === "safe" || oldBlockTag === "finalized") {
+      this._checkPostMergeBlockTags(oldBlockTag);
+
+      return this._node.getLatestBlockNumber();
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts:1371]: Not obvious what we should do here, but I guess this is a good start.

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts [hunk 4]
```diff
@@ -1738,10 +1749,16 @@ Hardhat Network's forking functionality only works with blocks from at least spu
...
       coinbase: this.getCoinbaseAddress(),
-      nonce: "0x0000000000000042",
+      nonce: this.isPostMergeHardfork()
+        ? "0x0000000000000000"
+        : "0x0000000000000042",
       timestamp: blockTimestamp,
     };
 
+    if (this.isPostMergeHardfork()) {
+      headerData.mixHash = this._getNextMixHash();
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/hardhat-network/provider/node.ts:1759]: I think we should always call `_getNextMixHash` and place this HF check inside it.

The logic is wrong though, mixHash is 0 post merge according to https://eips.ethereum.org/EIPS/eip-3675

Comment (alcuadrado) [reply packages/hardhat-core/src/internal/hardhat-network/provider/node.ts:1759]: Oh this other Paris overrides that I guess https://eips.ethereum.org/EIPS/eip-4399

**Comment by alcuadrado** [comment]
LGTM, except maybe I thought we were implementing an RPC method (and its helper) to manipulate the randao value. `hardhat_setPrevRandao(quantity)`.

File: docs/src/content/hardhat-network/docs/reference/index.md (hunk 2 above)
Comment (fvictorio) [reply docs/src/content/hardhat-network/docs/reference/index.md:65]: I added support for gray glacier in other PR: https://github.com/NomicFoundation/hardhat/pull/3087/commits/814a8e107dbeed6e027f5e1402fbc81af366924d

> Also, the new default should be merge.

I thought we were going to change the default after the mainnet merge, didn't we?

**Comment by fvictorio** [comment]
I added the `hardhat_setPrevRandao` method in a separate PR: https://github.com/NomicFoundation/hardhat/pull/3088

File: docs/src/content/hardhat-network/docs/reference/index.md (hunk 2 above)
Comment (alcuadrado) [reply docs/src/content/hardhat-network/docs/reference/index.md:65]: The default will be changed after the merge.

**Pull request merged**